File: backend/src/services/adminService.js
// src/services/adminService.js
// Administração global (Admin do Sistema): permissões (bloquear/desbloquear
// qualquer utilizador), convite de assessores e gestão de atividades de tudo
// o que acontece no sistema.
const bcrypt = require('bcryptjs');
const crypto = require('crypto');
const prisma = require('../config/database');
const {
  BusinessRuleError, NotFoundError, ConflictError, ValidationError,
} = require('../utils/errors');
const passwordPolicy = require('../utils/passwordPolicy');
const notificationService = require('./notificationService');
const config = require('../config/env');
const { AREAS_ADMIN, AREAS_ADMIN_LABEL } = require('../utils/adminAreas');

const pretty = (s) => String(s || '').replaceAll('_', ' ').toLowerCase();

async function listUsers() {
  const users = await prisma.user.findMany({
    select: {
      id: true, name: true, email: true, role: true, active: true, companyId: true, createdAt: true, adminAreas: true, company: { select: { name: true } },
    },
    orderBy: [{ active: 'asc' }, { createdAt: 'desc' }],
  });
  return users.map((u) => ({
    id: u.id, name: u.name, email: u.email, role: u.role, active: u.active, companyName: u.company?.name || null, createdAt: u.createdAt, adminAreas: u.adminAreas,
  }));
}

// Bloquear/desbloquear qualquer utilizador do sistema.
async function setUserStatus({ id, active, actingUserId }) {
  const user = await prisma.user.findUnique({ where: { id } });
  if (!user) throw new NotFoundError('Utilizador');
  if (id === actingUserId) throw new BusinessRuleError('Não pode bloquear a própria conta.');
  return prisma.user.update({ where: { id }, data: { active: Boolean(active) }, select: { id: true, name: true, active: true, role: true } });
}

// Atribuir áreas a um Admin do Sistema (vazio = Super Admin).
//
// Não pode alterar a PRÓPRIA conta: um Super Admin que restringisse as suas
// próprias áreas por engano ficava de fora das rotas que só o Super Admin
// gere — incluindo esta, que devolveria o próprio acesso. Sem outro Super
// Admin para o desfazer, é um bloqueio sem saída.
async function setUserAreas({ id, areas, actingUserId }) {
  const user = await prisma.user.findUnique({ where: { id } });
  if (!user) throw new NotFoundError('Utilizador');
  if (user.role !== 'ADMIN_SISTEMA') throw new BusinessRuleError('Áreas só se aplicam a Admin do Sistema.');
  if (id === actingUserId) throw new BusinessRuleError('Não pode alterar as próprias áreas.');
  return prisma.user.update({
    where: { id }, data: { adminAreas: areas }, select: { id: true, name: true, role: true, adminAreas: true },
  });
}

// ---------------------------------------------------------------------------
// Convite de Assessor (ADMIN_SISTEMA por área)
//
// Reutiliza a MESMA tabela e o MESMO ciclo de vida do convite de funcionário
// (EmployeeInvite: Pendente/Aceite/Expirado/Cancelado, reenvio, cancelamento)
// — não há uma segunda tabela nem um segundo sistema de convites.
//
// O TOKEN é diferente de propósito. O convite de funcionário assina um JWT
// com o papel embutido (authService.signInvite) — funciona porque a única
// coisa em jogo é uma conta de empresa. Aqui a aposta é maior: aceitar este
// convite cria um utilizador com poder sobre a PLATAFORMA INTEIRA. Por isso o
// token é um valor aleatório opaco (32 bytes, crypto.randomBytes) que não
// guarda NADA lá dentro — é só a chave de consulta. As áreas, o nome e o
// email vêm sempre da linha na base de dados, nunca do token, e nunca do que
// o assessor envia no pedido de aceitação.
// ---------------------------------------------------------------------------

const ADMIN_INVITE_TTL_DIAS = 7;

function gerarTokenDeConvite() {
  return crypto.randomBytes(32).toString('hex');
}

function buildAdminInviteEmail({ name, link, areaLabels }) {
  const subject = 'Convite para acesso administrativo ao Kixima';
  const listaAreas = areaLabels.join(', ');
  const text = [
    `Olá ${name},`, '',
    'Foi convidado a ser Admin do Sistema KIXIMA, com acesso às seguintes áreas:',
    listaAreas, '',
    'Clique no link abaixo para completar o seu cadastro e definir a sua senha:', link, '',
    `Este link é válido por ${ADMIN_INVITE_TTL_DIAS} dias e só pode ser usado uma vez.`, '',
    'Se não esperava este convite, ignore este email — ninguém consegue aceder com ele sem o clicar.', '',
    'Equipa Kixima.',
  ].join('\n');
  const html = `
    <div style="font-family:Arial,Helvetica,sans-serif;color:#1a1a1a;line-height:1.5">
      <p>Olá <strong>${name}</strong>,</p>
      <p>Foi convidado a ser <strong>Admin do Sistema KIXIMA</strong>, com acesso às seguintes áreas:</p>
      <p><strong>${listaAreas}</strong></p>
      <p>Clique no botão abaixo para completar o seu cadastro e definir a sua senha:</p>
      <p style="margin:22px 0">
        <a href="${link}" style="background:#c1121f;color:#fff;text-decoration:none;padding:12px 22px;border-radius:8px;font-weight:600;display:inline-block">Completar Cadastro</a>
      </p>
      <p style="font-size:13px;color:#666">Este link é válido por ${ADMIN_INVITE_TTL_DIAS} dias e só pode ser usado uma vez.</p>
      <p style="font-size:13px;color:#666">Se não esperava este convite, ignore este email.</p>
      <p>Equipa Kixima.</p>
    </div>`;
  return { subject, text, html };
}

async function sendAdminInviteEmail(invite, baseUrl = null) {
  const base = String(process.env.APP_URL || baseUrl || config.appUrl || '').replace(/\/$/, '');
  const link = `${base}/convite-admin/${invite.token}`;
  const areaLabels = (invite.adminAreas || []).map((a) => AREAS_ADMIN_LABEL[a] || a);
  const { subject, text, html } = buildAdminInviteEmail({ name: invite.name, link, areaLabels });
  await notificationService.sendEmail(invite.email, subject, text, { html });
  return link;
}

function applyExpiry(invite) {
  if (invite.status === 'PENDENTE' && invite.expiresAt && invite.expiresAt.getTime() < Date.now()) {
    return { ...invite, status: 'EXPIRADO' };
  }
  return invite;
}

function pickAdminInvite(i) {
  return {
    id: i.id, name: i.name, email: i.email, adminAreas: i.adminAreas, status: i.status,
    expiresAt: i.expiresAt, acceptedAt: i.acceptedAt, createdAt: i.createdAt,
  };
}

// Convidar um novo assessor. Só chega aqui quem já passou por
// requireSuperAdmin() na rota — mas a validação das áreas fica aqui também,
// porque um serviço não deve confiar em quem o chama para lhe poupar o
// trabalho de se defender sozinho.
async function createAdminInvite({ name, email, adminAreas }, createdById = null, baseUrl = null) {
  const areas = [...new Set(Array.isArray(adminAreas) ? adminAreas : [])];
  const invalidas = areas.filter((a) => !AREAS_ADMIN.includes(a));
  if (invalidas.length) throw new BusinessRuleError(`Área desconhecida: ${invalidas.join(', ')}.`);
  // VAZIO nunca se grava num convite: promoveria a Super Admin sem ninguém
  // ter decidido isso explicitamente através do ecrã de áreas.
  if (!areas.length) throw new BusinessRuleError('Selecione pelo menos uma área administrativa.');

  const normEmail = String(email).trim().toLowerCase();
  if (await prisma.user.findUnique({ where: { email: normEmail } })) {
    throw new ConflictError('Já existe uma conta com este email.');
  }

  const token = gerarTokenDeConvite();
  const expiresAt = new Date(Date.now() + ADMIN_INVITE_TTL_DIAS * 24 * 60 * 60 * 1000);
  const invite = await prisma.employeeInvite.create({
    data: {
      companyId: null, name: String(name).trim(), email: normEmail, role: 'ADMIN_SISTEMA',
      adminAreas: areas, token, status: 'PENDENTE', expiresAt, createdById,
    },
  });
  await sendAdminInviteEmail(invite, baseUrl);
  return pickAdminInvite(invite);
}

async function listAdminInvites() {
  const invites = await prisma.employeeInvite.findMany({
    where: { role: 'ADMIN_SISTEMA' }, orderBy: { createdAt: 'desc' },
  });
  return invites.map((i) => pickAdminInvite(applyExpiry(i)));
}

async function getAdminInvite(inviteId) {
  const invite = await prisma.employeeInvite.findUnique({ where: { id: inviteId } });
  if (!invite || invite.role !== 'ADMIN_SISTEMA') throw new NotFoundError('Convite');
  return invite;
}

// Reenvia: gera um token NOVO (o anterior fica órfão — já não corresponde a
// nada, mesmo que alguém ainda o tivesse guardado) e reabre o prazo.
async function resendAdminInvite(inviteId, baseUrl = null) {
  const invite = await getAdminInvite(inviteId);
  if (invite.status === 'ACEITO') throw new BusinessRuleError('Este convite já foi aceite.');
  const token = gerarTokenDeConvite();
  const expiresAt = new Date(Date.now() + ADMIN_INVITE_TTL_DIAS * 24 * 60 * 60 * 1000);
  const updated = await prisma.employeeInvite.update({
    where: { id: inviteId }, data: { token, status: 'PENDENTE', expiresAt, acceptedAt: null },
  });
  await sendAdminInviteEmail(updated, baseUrl);
  return pickAdminInvite(updated);
}

async function cancelAdminInvite(inviteId) {
  const invite = await getAdminInvite(inviteId);
  if (invite.status === 'ACEITO') throw new BusinessRuleError('Este convite já foi aceite.');
  const updated = await prisma.employeeInvite.update({ where: { id: inviteId }, data: { status: 'CANCELADO' } });
  return pickAdminInvite(updated);
}

/**
 * Valida o token de um convite de assessor — por CONSULTA à base, não por
 * verificação de assinatura: o próprio token não guarda nada, é só a chave.
 * Só uma linha PENDENTE e dentro do prazo é válida; qualquer outro estado
 * (cancelado, já aceite, expirado) é o mesmo "não" com uma explicação
 * diferente, para quem vê o ecrã perceber porquê.
 */
async function getInviteForAdminToken(token) {
  const invite = await prisma.employeeInvite.findUnique({ where: { token: String(token || '') } });
  if (!invite || invite.role !== 'ADMIN_SISTEMA') throw new BusinessRuleError('Convite inválido.');
  if (invite.status === 'CANCELADO') throw new BusinessRuleError('Este convite foi cancelado.');
  if (invite.status === 'ACEITO') throw new BusinessRuleError('Este convite já foi utilizado.');
  if (invite.expiresAt && invite.expiresAt.getTime() < Date.now()) throw new BusinessRuleError('Este convite expirou.');
  return invite;
}

// Resolve um convite (público) — o que o ecrã de aceitação mostra antes do
// assessor definir a senha. As áreas vêm só para MOSTRAR; nunca são lidas de
// volta a partir do que o assessor submeter a seguir.
async function resolveAdminInvite(token) {
  const invite = await getInviteForAdminToken(token);
  return {
    name: invite.name, email: invite.email, adminAreas: invite.adminAreas,
  };
}

// Aceitação (pública): o assessor define a senha e a conta é criada já ATIVA
// — ao contrário do convite de funcionário, não há um segundo aprovador
// depois disto; o Super Admin já decidiu ao enviar o convite.
//
// Note-se a assinatura: só recebe `password`. Não há parâmetro `adminAreas`
// nem `name`/`email` vindos do pedido — mesmo que o corpo do pedido HTTP
// trouxesse esses campos, esta função nunca os lê. As áreas, o nome e o email
// são sempre os da linha do convite.
async function acceptAdminInvite(token, { password }) {
  const invite = await getInviteForAdminToken(token);
  if (await prisma.user.findUnique({ where: { email: invite.email } })) {
    throw new ConflictError('Já existe uma conta com este email.');
  }
  const erroSenha = passwordPolicy.validar(password, { role: 'ADMIN_SISTEMA', email: invite.email });
  if (erroSenha) throw new ValidationError(erroSenha);

  const passwordHash = await bcrypt.hash(password, 12);
  // Transação: criar a conta e invalidar o convite têm de acontecer as duas
  // ou nenhuma. Sem isto, uma falha a meio deixava um convite "PENDENTE" que
  // na prática já tinha sido usado — reabrindo a porta para o token ser
  // reutilizado, exatamente o que "utilização única" promete impedir.
  return prisma.$transaction(async (tx) => {
    const user = await tx.user.create({
      data: {
        name: invite.name, email: invite.email, passwordHash, role: 'ADMIN_SISTEMA',
        companyId: null, adminAreas: invite.adminAreas, active: true, termsAcceptedAt: new Date(),
      },
      select: {
        id: true, name: true, email: true, role: true, adminAreas: true, active: true, createdAt: true,
      },
    });
    await tx.employeeInvite.update({ where: { id: invite.id }, data: { status: 'ACEITO', acceptedAt: new Date() } });
    return user;
  });
}

// Feed de atividades de todo o sistema.
async function systemActivities() {
  const [orders, companies, users, payments, tickets, policies, kpis] = await Promise.all([
    prisma.purchaseOrder.findMany({ include: { buyerCompany: { select: { name: true } }, supplierCompany: { select: { name: true } } }, orderBy: { updatedAt: 'desc' }, take: 25 }),
    prisma.company.findMany({ orderBy: { createdAt: 'desc' }, take: 10 }),
    prisma.user.findMany({ orderBy: { createdAt: 'desc' }, take: 10, include: { company: { select: { name: true } } } }),
    prisma.payment.findMany({ orderBy: { processedAt: 'desc' }, take: 10 }),
    prisma.supportTicket.findMany({ orderBy: { createdAt: 'desc' }, take: 10 }),
    prisma.supplierToKiximaPolicy.findMany({ orderBy: { createdAt: 'desc' }, take: 10, include: { company: { select: { name: true } } } }),
    (async () => ({
      empresas: await prisma.company.count(),
      utilizadores: await prisma.user.count(),
      ordens: await prisma.purchaseOrder.count(),
      pagamentos: await prisma.payment.count(),
      tickets: await prisma.supportTicket.count(),
    }))(),
  ]);

  const ev = [];
  for (const o of orders) ev.push({ type: 'Ordem de Compra', module: 'Compras', title: `PO ${o.reference}`, detail: `${pretty(o.status)} — ${o.buyerCompany?.name} → ${o.supplierCompany?.name}`, at: o.updatedAt });
  for (const c of companies) ev.push({ type: 'Empresa', module: 'Credenciamento', title: c.name, detail: `Empresa ${pretty(c.type)} — estado ${pretty(c.status)}`, at: c.createdAt });
  for (const u of users) ev.push({ type: 'Utilizador', module: 'Contas', title: u.name, detail: `Novo ${pretty(u.role)} — ${u.company?.name || '—'}`, at: u.createdAt });
  for (const p of payments) ev.push({ type: 'Pagamento', module: 'Financeiro', title: p.reference, detail: 'Pagamento processado', at: p.processedAt });
  for (const t of tickets) ev.push({ type: 'Suporte', module: 'Ajuda', title: `#${t.reference}`, detail: t.subject, at: t.createdAt });
  for (const pl of policies) ev.push({ type: 'Apólice', module: 'Seguros', title: pl.policyNumber, detail: `${pl.insurer} — ${pl.company?.name || ''}`, at: pl.createdAt });
  ev.sort((a, b) => new Date(b.at) - new Date(a.at));

  return { kpis, items: ev.slice(0, 60) };
}

// Livro de taxas da plataforma (KIXIMA) — todas as taxas geradas nos pagamentos.
async function listPlatformFees() {
  const fees = await prisma.platformFee.findMany({
    orderBy: { createdAt: 'desc' },
    include: {
      company: { select: { name: true, type: true } },
      invoice: { select: { reference: true, amount: true, currency: true } },
    },
  });
  const totalAOA = fees.reduce((s, f) => s + Number(f.amount), 0);
  const pendingAOA = fees.filter((f) => f.status === 'PENDENTE').reduce((s, f) => s + Number(f.amount), 0);
  return {
    fees,
    kpis: { total: fees.length, totalAOA, pendingAOA, cobradas: fees.filter((f) => f.status === 'COBRADO').length },
  };
}

// Marcar uma taxa como cobrada (enquanto não há débito automático).
async function chargePlatformFee(id) {
  const fee = await prisma.platformFee.findUnique({ where: { id } });
  if (!fee) throw new NotFoundError('Taxa');
  return prisma.platformFee.update({ where: { id }, data: { status: 'COBRADO', chargedAt: new Date() } });
}

module.exports = {
  listUsers,
  setUserStatus,
  setUserAreas,
  createAdminInvite,
  listAdminInvites,
  resendAdminInvite,
  cancelAdminInvite,
  resolveAdminInvite,
  acceptAdminInvite,
  systemActivities,
  listPlatformFees,
  chargePlatformFee,
};

File: backend-java/src/main/java/ao/kixima/admin/AdminService.java
package ao.kixima.admin;

import ao.kixima.admin.dto.AdminInviteDto;
import ao.kixima.admin.dto.AdminUserDto;
import ao.kixima.admin.dto.ResolvedAdminInviteDto;
import ao.kixima.common.error.BusinessRuleException;
import ao.kixima.common.error.ConflictException;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.common.error.ValidationException;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.company.SupplierToKiximaPolicy;
import ao.kixima.company.SupplierToKiximaPolicyRepository;
import ao.kixima.invite.EmployeeInvite;
import ao.kixima.invite.EmployeeInviteRepository;
import ao.kixima.invite.InviteStatus;
import ao.kixima.notification.EmailDispatchService;
import ao.kixima.payment.Payment;
import ao.kixima.payment.PaymentRepository;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.po.PurchaseOrderRepository;
import ao.kixima.security.AdminArea;
import ao.kixima.security.PersonaRole;
import ao.kixima.support.SupportTicket;
import ao.kixima.support.SupportTicketRepository;
import ao.kixima.user.PasswordPolicy;
import ao.kixima.user.User;
import ao.kixima.user.UserRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.PageRequest;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Espelha backend/src/services/adminService.js — utilizadores da plataforma
 * (bloquear/desbloquear, áreas), convites de assessor ADMIN_SISTEMA e o feed
 * de actividades de todo o sistema. As taxas da plataforma vivem em
 * {@link ao.kixima.payment.PlatformFeeService} (Lacunas A.3).
 *
 * O convite de assessor reutiliza a MESMA tabela e ciclo de vida do convite
 * de funcionário ({@link EmployeeInvite}); o que muda é o token — aleatório,
 * opaco, sem nada lá dentro — e o facto de nunca se confiar em áreas vindas
 * do pedido do próprio assessor.
 */
@Service
public class AdminService {

    static final int ADMIN_INVITE_TTL_DIAS = 7;
    private static final SecureRandom RANDOM = new SecureRandom();

    private final UserRepository userRepository;
    private final EmployeeInviteRepository inviteRepository;
    private final CompanyRepository companyRepository;
    private final PurchaseOrderRepository purchaseOrderRepository;
    private final PaymentRepository paymentRepository;
    private final SupportTicketRepository supportTicketRepository;
    private final SupplierToKiximaPolicyRepository policyRepository;
    private final PasswordEncoder passwordEncoder;
    private final PasswordPolicy passwordPolicy;
    private final EmailDispatchService emailDispatchService;
    private final String appUrl;

    public AdminService(UserRepository userRepository, EmployeeInviteRepository inviteRepository,
                        CompanyRepository companyRepository, PurchaseOrderRepository purchaseOrderRepository,
                        PaymentRepository paymentRepository, SupportTicketRepository supportTicketRepository,
                        SupplierToKiximaPolicyRepository policyRepository, PasswordEncoder passwordEncoder,
                        PasswordPolicy passwordPolicy, EmailDispatchService emailDispatchService,
                        @Value("${kixima.app-url:}") String appUrl) {
        this.userRepository = userRepository;
        this.inviteRepository = inviteRepository;
        this.companyRepository = companyRepository;
        this.purchaseOrderRepository = purchaseOrderRepository;
        this.paymentRepository = paymentRepository;
        this.supportTicketRepository = supportTicketRepository;
        this.policyRepository = policyRepository;
        this.passwordEncoder = passwordEncoder;
        this.passwordPolicy = passwordPolicy;
        this.emailDispatchService = emailDispatchService;
        this.appUrl = appUrl == null ? "" : appUrl;
    }

    private static String pretty(Object s) {
        return s == null ? "" : String.valueOf(s).replace("_", " ").toLowerCase();
    }

    // --- Utilizadores ----------------------------------------------------

    @Transactional(readOnly = true)
    public List<AdminUserDto> listUsers() {
        return userRepository.findTodosOrderByActiveAscCreatedAtDesc().stream()
                .map(u -> new AdminUserDto(u.getId(), u.getName(), u.getEmail(), u.getRole().name(), u.isActive(),
                        u.getCompany() == null ? null : u.getCompany().getName(), u.getCreatedAt(), u.getAdminAreas()))
                .toList();
    }

    @Transactional
    public Map<String, Object> setUserStatus(String id, boolean active, String actingUserId) {
        User user = userRepository.findById(id).orElseThrow(() -> new NotFoundException("Utilizador"));
        if (id.equals(actingUserId)) throw new BusinessRuleException("Não pode bloquear a própria conta.");
        user.setActive(active);
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", user.getId());
        m.put("name", user.getName());
        m.put("active", user.isActive());
        m.put("role", user.getRole().name());
        return m;
    }

    @Transactional
    public Map<String, Object> setUserAreas(String id, List<String> areas, String actingUserId) {
        User user = userRepository.findById(id).orElseThrow(() -> new NotFoundException("Utilizador"));
        if (user.getRole() != PersonaRole.ADMIN_SISTEMA) throw new BusinessRuleException("Áreas só se aplicam a Admin do Sistema.");
        if (id.equals(actingUserId)) throw new BusinessRuleException("Não pode alterar as próprias áreas.");
        user.setAdminAreas(new ArrayList<>(areas));
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", user.getId());
        m.put("name", user.getName());
        m.put("role", user.getRole().name());
        m.put("adminAreas", user.getAdminAreas());
        return m;
    }

    // --- Convites de assessor -------------------------------------------

    /** 32 bytes aleatórios em hex — nada assinado, nada dentro: a base é a única fonte de verdade. */
    static String gerarTokenDeConvite() {
        byte[] b = new byte[32];
        RANDOM.nextBytes(b);
        return HexFormat.of().formatHex(b);
    }

    private static List<String> areasValidadas(List<String> adminAreas) {
        List<String> areas = new ArrayList<>(new LinkedHashSet<>(adminAreas == null ? List.of() : adminAreas));
        List<String> invalidas = areas.stream().filter(a -> !AdminArea.AREAS_ADMIN.contains(a)).toList();
        if (!invalidas.isEmpty()) throw new BusinessRuleException("Área desconhecida: " + String.join(", ", invalidas) + ".");
        if (areas.isEmpty()) throw new BusinessRuleException("Selecione pelo menos uma área administrativa.");
        return areas;
    }

    @Transactional
    public AdminInviteDto createAdminInvite(String name, String email, List<String> adminAreas, String createdById, String baseUrl) {
        List<String> areas = areasValidadas(adminAreas);
        String normEmail = email == null ? "" : email.trim().toLowerCase();
        if (userRepository.findByEmail(normEmail).isPresent()) throw new ConflictException("Já existe uma conta com este email.");
        Instant agora = Instant.now();
        EmployeeInvite invite = new EmployeeInvite(UUID.randomUUID().toString(), null, name.trim(), normEmail,
                PersonaRole.ADMIN_SISTEMA, gerarTokenDeConvite(), agora.plus(Duration.ofDays(ADMIN_INVITE_TTL_DIAS)), createdById, agora);
        invite.setAdminAreas(areas);
        inviteRepository.save(invite);
        sendAdminInviteEmail(invite, baseUrl);
        return AdminInviteDto.de(invite);
    }

    @Transactional(readOnly = true)
    public List<AdminInviteDto> listAdminInvites() {
        return inviteRepository.findByRoleOrderByCreatedAtDesc(PersonaRole.ADMIN_SISTEMA).stream()
                .map(AdminInviteDto::comExpiracao).toList();
    }

    private EmployeeInvite getAdminInvite(String inviteId) {
        EmployeeInvite invite = inviteRepository.findById(inviteId).orElse(null);
        if (invite == null || invite.getRole() != PersonaRole.ADMIN_SISTEMA) throw new NotFoundException("Convite");
        return invite;
    }

    @Transactional
    public AdminInviteDto resendAdminInvite(String inviteId, String baseUrl) {
        EmployeeInvite invite = getAdminInvite(inviteId);
        if (invite.getStatus() == InviteStatus.ACEITO) throw new BusinessRuleException("Este convite já foi aceite.");
        invite.setToken(gerarTokenDeConvite());
        invite.setStatus(InviteStatus.PENDENTE);
        invite.setExpiresAt(Instant.now().plus(Duration.ofDays(ADMIN_INVITE_TTL_DIAS)));
        invite.setAcceptedAt(null);
        sendAdminInviteEmail(invite, baseUrl);
        return AdminInviteDto.de(invite);
    }

    @Transactional
    public AdminInviteDto cancelAdminInvite(String inviteId) {
        EmployeeInvite invite = getAdminInvite(inviteId);
        if (invite.getStatus() == InviteStatus.ACEITO) throw new BusinessRuleException("Este convite já foi aceite.");
        invite.setStatus(InviteStatus.CANCELADO);
        return AdminInviteDto.de(invite);
    }

    /** Um token que não existe é um erro de negócio (400), não uma excepção crua nem um 404 que confirme o que existe. */
    private EmployeeInvite getInviteForAdminToken(String token) {
        EmployeeInvite invite = token == null || token.isBlank() ? null : inviteRepository.findByToken(token).orElse(null);
        if (invite == null || invite.getRole() != PersonaRole.ADMIN_SISTEMA) throw new BusinessRuleException("Convite inválido.");
        if (invite.getStatus() == InviteStatus.CANCELADO) throw new BusinessRuleException("Este convite foi cancelado.");
        if (invite.getStatus() == InviteStatus.ACEITO) throw new BusinessRuleException("Este convite já foi utilizado.");
        if (invite.getExpiresAt() != null && invite.getExpiresAt().isBefore(Instant.now())) throw new BusinessRuleException("Este convite expirou.");
        return invite;
    }

    @Transactional(readOnly = true)
    public ResolvedAdminInviteDto resolveAdminInvite(String token) {
        EmployeeInvite invite = getInviteForAdminToken(token);
        return new ResolvedAdminInviteDto(invite.getName(), invite.getEmail(), invite.getAdminAreas());
    }

    /**
     * Aceitação: a conta nasce ATIVA (o Super Admin já decidiu ao convidar) e
     * com as áreas DO CONVITE — nunca as do pedido, que o controller nem lê.
     */
    @Transactional
    public User acceptAdminInvite(String token, String password) {
        EmployeeInvite invite = getInviteForAdminToken(token);
        if (userRepository.findByEmail(invite.getEmail()).isPresent()) throw new ConflictException("Já existe uma conta com este email.");
        String erroSenha = passwordPolicy.validar(password, PersonaRole.ADMIN_SISTEMA, invite.getEmail());
        if (erroSenha != null) throw new ValidationException(erroSenha);
        Instant agora = Instant.now();
        User user = new User(UUID.randomUUID().toString(), invite.getName(), invite.getEmail(), passwordEncoder.encode(password),
                PersonaRole.ADMIN_SISTEMA, null, true, agora, agora);
        user.setAdminAreas(new ArrayList<>(invite.getAdminAreas()));
        userRepository.save(user);
        invite.setStatus(InviteStatus.ACEITO);
        invite.setAcceptedAt(agora);
        return user;
    }

    public static Map<String, Object> utilizadorAceite(User user) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", user.getId());
        m.put("name", user.getName());
        m.put("email", user.getEmail());
        m.put("role", user.getRole().name());
        m.put("adminAreas", user.getAdminAreas());
        m.put("active", user.isActive());
        m.put("createdAt", user.getCreatedAt());
        return m;
    }

    private void sendAdminInviteEmail(EmployeeInvite invite, String baseUrl) {
        String base = !appUrl.isBlank() ? appUrl : (baseUrl == null ? "" : baseUrl);
        String link = base.replaceAll("/$", "") + "/convite-admin/" + invite.getToken();
        String listaAreas = String.join(", ", invite.getAdminAreas().stream()
                .map(a -> AdminArea.AREAS_ADMIN_LABEL.getOrDefault(a, a)).toList());
        String texto = String.join("\n",
                "Olá " + invite.getName() + ",", "",
                "Foi convidado a ser Admin do Sistema KIXIMA, com acesso às seguintes áreas:",
                listaAreas, "",
                "Clique no link abaixo para completar o seu cadastro e definir a sua senha:", link, "",
                "Este link é válido por " + ADMIN_INVITE_TTL_DIAS + " dias e só pode ser usado uma vez.", "",
                "Se não esperava este convite, ignore este email — ninguém consegue aceder com ele sem o clicar.", "",
                "Equipa Kixima.");
        emailDispatchService.dispatch(invite.getEmail(), "Convite para acesso administrativo ao Kixima", texto);
    }

    // --- Gestão de Atividades -------------------------------------------

    @Transactional(readOnly = true)
    public Map<String, Object> systemActivities() {
        List<PurchaseOrder> orders = purchaseOrderRepository.findRecentesComEmpresas(PageRequest.of(0, 25));
        List<Company> companies = companyRepository.findTop10ByOrderByCreatedAtDesc();
        List<User> users = userRepository.findTop10ByOrderByCreatedAtDesc();
        List<Payment> payments = paymentRepository.findTop10ByOrderByProcessedAtDesc();
        List<SupportTicket> tickets = supportTicketRepository.findTop10ByOrderByCreatedAtDesc();
        List<SupplierToKiximaPolicy> policies = policyRepository.findTop10ByOrderByCreatedAtDesc();

        Map<String, Object> kpis = new LinkedHashMap<>();
        kpis.put("empresas", companyRepository.count());
        kpis.put("utilizadores", userRepository.count());
        kpis.put("ordens", purchaseOrderRepository.count());
        kpis.put("pagamentos", paymentRepository.count());
        kpis.put("tickets", supportTicketRepository.count());

        List<Map<String, Object>> ev = new ArrayList<>();
        for (PurchaseOrder o : orders) {
            ev.add(evento("Ordem de Compra", "Compras", "PO " + o.getReference(),
                    pretty(o.getStatus()) + " — " + nome(o.getBuyerCompany()) + " → " + nome(o.getSupplierCompany()), o.getUpdatedAt()));
        }
        for (Company c : companies) {
            ev.add(evento("Empresa", "Credenciamento", c.getName(), "Empresa " + pretty(c.getType()) + " — estado " + pretty(c.getStatus()), c.getCreatedAt()));
        }
        for (User u : users) {
            Company empresa = u.getCompanyId() == null ? null : companyRepository.findById(u.getCompanyId()).orElse(null);
            ev.add(evento("Utilizador", "Contas", u.getName(), "Novo " + pretty(u.getRole()) + " — " + (empresa == null ? "—" : empresa.getName()), u.getCreatedAt()));
        }
        for (Payment p : payments) ev.add(evento("Pagamento", "Financeiro", p.getReference(), "Pagamento processado", p.getProcessedAt()));
        for (SupportTicket t : tickets) ev.add(evento("Suporte", "Ajuda", "#" + t.getReference(), t.getSubject(), t.getCreatedAt()));
        for (SupplierToKiximaPolicy pl : policies) {
            Company empresa = companyRepository.findById(pl.getCompanyId()).orElse(null);
            ev.add(evento("Apólice", "Seguros", pl.getPolicyNumber(), pl.getInsurer() + " — " + (empresa == null ? "" : empresa.getName()), pl.getCreatedAt()));
        }
        ev.sort((a, b) -> {
            Instant ia = (Instant) a.get("at");
            Instant ib = (Instant) b.get("at");
            if (ia == null && ib == null) return 0;
            if (ia == null) return 1;
            if (ib == null) return -1;
            return ib.compareTo(ia);
        });
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kpis", kpis);
        out.put("items", ev.subList(0, Math.min(60, ev.size())));
        return out;
    }

    private static String nome(Company c) {
        return c == null ? null : c.getName();
    }

    private static Map<String, Object> evento(String type, String module, String title, String detail, Instant at) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("type", type);
        m.put("module", module);
        m.put("title", title);
        m.put("detail", detail);
        m.put("at", at);
        return m;
    }
}
